### PR TITLE
Add infrastructure as code for record thumbnail lambda

### PR DESCRIPTION
**Pull request by liam-lloyd**
Because upload volume is highly variable, we would like to use a lambda to attach thumbnails created by Archivematica to new records in our database. This commit adds the necessary terraform configuration to deploy such a lambda, and updates the deploy actions of the repo to include the lambda in deploys. It also updates the lambda handler to correctly reflect the format of the messages that will trigger the lambda.

### CODE REVIEW

**Review by meisekimiu — APPROVED** [review]

File: .github/workflows/staging_deploy.yml
Comment (slifty) [inline .github/workflows/staging_deploy.yml:58]: There is a lot happening in one line here -- would it be possible to write this in a way that visually organizes the granular steps (even if it's all one command in the end)

File: Dockerfile.record_thumbnail_attacher
Comment (slifty) [inline Dockerfile.record_thumbnail_attacher:24]: What was the reason for removing this from the build step?  Depending on the answer it might be worth including in the commit description (or, potentially, having its own commit to explain the reasoning!)

File: packages/record_thumbnail_attacher/src/index.test.ts
Comment (slifty) [inline packages/record_thumbnail_attacher/src/index.test.ts:51]: Are these test change related to the infra as code changes or is it fixing a separate issue?

It may be worth putting them in their own commit just to contextualize the reason for the change.

File: terraform/test_cluster/variables.tf
Comment (slifty) [inline terraform/test_cluster/variables.tf:278]: I am *not* a terraform expert by any means -- do you know if we need to repeat these variable definitions across test / prod cluster, or might there be a pattern for sharing this kind of variable that is universal to all environments / clusters.

**Review by slifty — CHANGES_REQUESTED** [review]
I raised a few bigger-picture questions that would be great to talk through (and please note that the conclusion of that discussion can very well be "nah!')

Marking this as "request changes" just because it would be good to have that discussion before hitting the merge button, even if we end up deciding not to implement changes to this PR directly.

**Comment by liam-lloyd** [comment]
> I also get the sense that there is more than one commit's worth of work in this PR (but may very well be wrong wrong wrong). For instance, the changes that wrap various calls with a Record seems like it might be unrelated to the terraform work? If that's the case, let's split it into multiple commits with appropriate descriptions to help out with the git history.

The message format changes are indeed unrelated to the terraform work; I've divided them into separate commits.

File: .github/workflows/dev_deploy.yml
Comment (liam-lloyd) [reply .github/workflows/dev_deploy.yml:54]: I do think we want each lambda to have a separate build step; the O(n) pipeline steps don't worry me since the build steps can (and do) run in parallel. As long as each deployable object in this repo gets its own image, this seems like the way to do things.
In considering this, though, I have come to the realization the *reason* each deployable object in this repo gets its own image is a misunderstanding on my part. The first time I started working on a lambda in this project (about a year ago) I read that the code of a lambda had a size limit of 50 MB. I was concerned that in Typescript a monolithic project might exceed this, so I divided `stela` into multiple packages that could be deployed separately using npm workspaces. As it turns out though, that 50 MB limit only applies if you upload your code via zip file; if you use docker images they can be up to 10 GB. So while I originally though that a single application capable or running any of our lambdas or our api server or anything else in the project was off the table, I think I was mistaken. That said, the decoupling benefits of the current approach may be worthwhile on their own merits.

File: .github/workflows/dev_deploy.yml
Comment (liam-lloyd) [reply .github/workflows/dev_deploy.yml:80]: I agree it would be better to restrict what we build/deploy to what was actually changed. If I understand correctly, the path filter only works at the workflow level though, so we'd need to create separate workflows for each buildable thing. I don't think there's anything wrong with that, but it probably fits into the refactor suggested in the next comment more than it fits into this PR.

File: .github/workflows/full_test_deploy.yml
Comment (liam-lloyd) [reply .github/workflows/full_test_deploy.yml:51]: This issue occurred to me while I was working on this, and I created a [ticket](https://permanent.atlassian.net/browse/PER-9799) to at least turn the build steps into a reusable workflow. I think the environments + reusable workflows approach is the right one to take here, and we are actually using both of those features in some places already (the tests workflow is reusable, and we use environments to add protections to prod deploys). I think this might be a good project for me to undertake in the potholes sprint in the beginning of next month.

File: .github/workflows/dev_deploy.yml
Comment (slifty) [reply .github/workflows/dev_deploy.yml:56]: Fully agree with all of that analysis!

File: .github/workflows/staging_deploy.yml
Comment (liam-lloyd) [reply .github/workflows/staging_deploy.yml]: This is now broken into several pieces, which should make things clearer. Also, image tag generation is now a reusable workflow.

File: Dockerfile.record_thumbnail_attacher
Comment (liam-lloyd) [reply Dockerfile.record_thumbnail_attacher]: I pulled this out when I was trying to figure out what I thought was a dependency versioning issue. That issue turned out to be something else, so I've put this back in.

File: terraform/test_cluster/variables.tf
Comment (liam-lloyd) [reply terraform/test_cluster/variables.tf:278]: I've looked a bit, and I haven't found a way to do this for the variables files where the variables are declared. There is a way to do this with the variable _assignments_ in Terraform Cloud, but even that has limited utility for us, because of the way our test environments are in the same workspace (so usually there aren't duplicate variables across workspaces, the test one will have `dev_cloudfront_url` and `staging_cloudfront_url`, while the prod one will just have `cloudfront_url`). The reason the workspaces are organized this way is so we can save money by running staging and dev on the same EKS cluster.

File: .github/workflows/generate_image_tags.yml
Comment (slifty) [inline .github/workflows/generate_image_tags.yml:18]: This whole section is spectacular and self documenting, very nice!

**Review by slifty — APPROVED** [review]
This is really great work; and thank you for all of the thoughtful iteration!

I should note that I did not *run* any of this and this is a review based on code review rather than testing, but it's all looking good.

**Pull request merged**
